### DISCUSSION TITLE
[Validator] use early return instead of throwing an exception

### DISCUSSION THREAD

**Discussion by nemanjajojic** [symfony/symfony#45000] [Q&A]
## Description
I’m using validator as a standalone package and I’m run into something which imho is a bug, so I'd like to hear your opinions before I create a PR.

in `RecursiveContextualValidator.php` line `660` if passed value is not an object exception(`NoSuchMetadataException`) will be thrown. My question is "Why not use early return instead of throwing an exception?"

Example

Imagine we have an entity with propery `product` which is embedded Product entity, with following validation rules(atm Product validation rules are not important).
```
->addPropertyConstraints('product', [
                new Type(Product::class),
                new NotNull(),
                new Valid(),
            ])
```
if user pass anything which is not an object eg. int or string exception will be thrown instead of getting validation error(it’s API call with json payload :point_down:).

```
JSON request payload
{
   "product": 1
}
```

On the other hand, if we use early return instead of throwing an exception, we will see type validation error which is far more better imo.

## Proposal 
```
if (!\is_object($value)) {
            return;
        }
```
instead of
```
if (!\is_object($value)) {
            throw new NoSuchMetadataException(sprintf('Cannot create metadata for non-objects. Got: "%s".', \gettype($value)));
        }
```

thank you in advance

**Comment by healerz** [comment]
An alternative could be to build a violation before the early return to avoid the necessity of adding the `Type` constraint additionally to the `Valid` constraint and to prevent "silent" validation errors.

**Comment by Nek-** [comment]
I'm not sure you're using the validator correctly. Maybe you should share some code?

The following configuration works well (and you can obviously use it nested in another configuration):

```php
use Symfony\Component\Validator\Validation;
use Symfony\Component\Validator\Constraints\Type;
$validator = Validation::createValidator();
$violations = $validator->validate(1, [
    new Type(Product::class)
]);
```

**Reply by nemanjajojic** [reply]
@Nek- thank you for your response. I'll try to give you more context so you could have a better understanding of a problem that we have.

Let's say we have two entities `Test` and `Product`

### Entities

```
class Test {
   /**
     * @var string
     */
    private $id;
    
    /**
     * @var Product
     */
    private $product;
}
```

```
class Product {
   /**
     * @var string
     */
    private $id;
    
    /**
     * @var string
     */
    private $name;
}
```

### Validation mapping for both of them

`Test` entity
```
class TestValidationMapping {
    public function configureMapping(ClassMetadata $metadata): void
    {
        $metadata
            ->addPropertyConstraints('id', [
                new Type('string'),
                new NotNull(),
                new NotBlank(),
                new Uuid(),
            ])
            ->addPropertyConstraints('product', [
                new Type(Product::class),
                new NotNull(),
                new Valid(),
            ])
}
```

`Product` entity
```
class ProductValidationMapping {
    public function configureMapping(ClassMetadata $metadata): void
    {
        $metadata
            ->addPropertyConstraints('id', [
                new Type('string'),
                new NotNull(),
                new NotBlank(),
                new Uuid(),
            ])
            ->addPropertyConstraints('name', [
                new Type('string'),
                new NotNull(),
                new Length(['max' => 255]),
            ])
}
```

**This is pseudo-code not working example**.
```
$testEntity = new Test();
$testEntity->setProduct($request->getAttribute('product')); 

$validator = Validation::createValidator();
$violations = $validator->validate($testEntity);
```
So if the user sends `product` with a value that is not a json object in the request, an exception will be thrown instead of a validation message, since the product is not valid. The reason for this lies in `RecursiveContextualValidator.php` line `660`.

And yes I know we could validate individual param from request first and then create entity object in valid state, but the project is already set up in the way we are validating entity and we don't wanna have validation rules at two places.

Let me know if you need more info.
